### ISSUE TITLE
Mypy says type parameter not defined in some cases of generic function definition where the function header doesn't use the type parameter

### ISSUE THREAD

**Issue by janramon** [python/mypy#19947]
**Bug Report**

Given the following code:

```python
class A:
    pass

class C[T]:
    pass

def f[T:A]() -> None:
    y = C[T]()    # Line 9: error
```

mypy 1.18.2 outputs the error 

```
$ mypy test.py
test.py:9: error: Name "T" is not defined  [name-defined]
Found 1 error in 1 file (checked 1 source file)
```

The code doesn't look wrong to me.  pyright doesn't output an error.  

I understand that none of the arguments to the function is annotated with a type, and so at runtime the type parameter T won't have any influence on the content of the function.  This is mainly due to the fact that type annotation syntax isn't sufficiently expressive to allow me to express the relevance of T to the input parameters in the specific / more complex application I'm working on (in particular with multiple type parameters, it is illegal to use on one type parameter a bound depending on another type parameter).  Nevertheless, even if not very useful the code seems legal.

**To Reproduce**

With or without the :A bound on T, the error is reported.
Adding an argument to f depending on T lets the error disappear:

```python
class A:
    pass

class C[T]:
    pass

def f[T:A](x:T|None=None) -> None:
    y = C[T]()    # No error
```

As the x:T argument is not really relevant to the legality or type-consistency of the code, its presence shouldn't influence whether or not an error is reported.

**Expected Behavior**

I guess the code is correct and in both above cases no error should be reported.

**Actual Behavior**

If there is no argument with type annotation somehow depending on T, an error is reported.  

**Your Environment**

- Mypy version used: 1.18.2
- Mypy command-line flags: None (the behavior is the same adding "--strict")
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.12.6

**Comment by neopexrl** [comment]
Hi, can I work on this issue?
